### ISSUE TITLE
Job queued but not run

### ISSUE THREAD

**Issue by fossilet** [nvie/rq#758]
I have a job not run, in the log of worker: 

2016-11-15 23:18:34,617 11894 {q}default: [job details] (e47b96dc-d0f8-48b4-bbc9-090e6443080a)
2016-11-15 23:18:34,618 11894 Sent heartbeat to prevent worker timeout. Next one should arrive within 420 seconds.
2016-11-15 23:18:34,621 22069 Sent heartbeat to prevent worker timeout. Next one should arrive within 420 seconds.
2016-11-15 23:18:40,741 11894 Sent heartbeat to prevent worker timeout. Next one should arrive within 420 seconds.
2016-11-15 23:18:40,743 11894 
2016-11-15 23:18:40,743 11894 *** Listening on {q}high, {q}default...
2016-11-15 23:18:40,743 11894 Sent heartbeat to prevent worker timeout. Next one should arrive within 420 seconds.

22069 is the forked process to run the job. There should have been lines like below or error:

[timestamp] 22069 Sent heartbeat to prevent worker timeout. Next one should arrive within 420 seconds.
[timestamp] 22069 {q}high: Job OK (e47b96dc-d0f8-48b4-bbc9-090e6443080a)
[timestamp] 22069 Result is kept for 500 seconds

In redis, the job status is queued:

> hgetall rq:job:e47b96dc-d0f8-48b4-bbc9-090e6443080a
 1) "created_at"
 2) "2016-11-15T15:18:34Z"
 3) "status"
 4) "queued"
...

This happens only occasionally.

**Comment by zealseeker** [comment]
I got the same problem. All the jobs I submitted were in a 'queued' status. using the [demo scripts](http://python-rq.org/docs/). If you use `job.perform()` and it will return the result indicating that the jobs is ok. But I don't know the the queue didn't "START TO WORK".

**Comment by selwin** [comment]
Are you sure the queue and worker use the same Redis connection? I'm not aware of any bug in RQ that could lead to a worker not seeing jobs that are queued properly.

**Comment by nolanbrown** [comment]
I've been having the same issue. The redis connection is setup properly, I've inspected everything but sometimes queued jobs stay queued. They don't timeout or ever get processed by the worker.

**Comment by ssikiki** [comment]
+1 the same problem

My env:  rq==0.8.0   redis==2.8.19

I've found that worker sometimes doesn't fetch job .
I'm sure the job has been sent to rq,  because when I enqueued a job,  I get the job status  immediately, and it shows 'queued'. 

taskid: 0b972f2e-48e7-4710-bb51-085a83978c56
queue: **host.6JGQ2Y1**
enqueued_at: 2017-06-02 09:14:55
status: **queued**        

And I'm sure the worker wasn't dead, because when I sent another job to the same queue and same worker, it can be done.

below is the log,  it shows the worker listen on two queues, one of them is **host.6JGQ2Y1**,  , is alive  from 6.1 20:02 to 6.2 11:21, but the job at 6.2 9:14 is missed

[2017-06-01 20:02:27] INFO [rq.worker:502] storage.ceph: consumer.storage.vol_act_v3()
[2017-06-01 20:02:27] INFO [rq.worker:99] storage.ceph: Job OK (6080387d-7779-4509-bd68-f18e917b4b7a)
[2017-06-01 20:02:27] INFO [rq.worker:108] Result is kept for 86400 seconds
[2017-06-01 20:02:27] INFO [rq.worker:489]
[2017-06-01 20:02:27] INFO [rq.worker:490] *** Listening on host.6JGQ2Y1, storage.ceph...
[2017-06-02 11:21:15] INFO [rq.worker:502] host.6JGQ2Y1: consumer.vm.nic_act_v3()
[2017-06-02 11:21:15] INFO [rq.worker:99] host.6JGQ2Y1: Job OK (b2d694c2-d49e-4b41-8a2c-2d0a6288e452)
[2017-06-02 11:21:15] INFO [rq.worker:108] Result is kept for 86400 seconds
[2017-06-02 11:21:16] INFO [rq.worker:802] Cleaning registries for queue: host.6JGQ2Y1
[2017-06-02 11:21:16] INFO [rq.worker:802] Cleaning registries for queue: storage.ceph

This issue doesn't happen very often,  so it's difficult to reproduce,  could someone give me some tips how to debug the problem ? (from rq version 0.6.x to 0.8.0 , the issue has occurred)

**Comment by ssikiki** [comment]
I've found that when instantiate Worker class, set default_worker_ttl to 120 (default is 420) , The problem never happened again

**Comment by hitigon** [comment]
Having the same issue with latest rq + redis-cluster 
@ssikiki
@nolanbrown 
any chances you guys are using redis cluster as well?

**Comment by hitigon** [comment]
@ssikiki interesting! I'll give a try

**Comment by hitigon** [comment]
changing default_worker_ttl to 120 didn't work for me.

**Comment by fossilet** [comment]
@hitigon I remember rq does not have native redis cluster support. Regarding data loss with redis-cluster, there are several other issues maybe of help to you.

**Comment by hitigon** [comment]
@fossilet yeah, I am trying to make rq work with Redis Cluster. I am using redis-py-cluster which is based on redis.py. There might be some data loss happening as I saw rpush/lpush a job id to the queue didn't happen sometimes. This also may be related to the implementation of pipeline in redis-py-cluster

**Comment by fossilet** [comment]
@hitigon I had some try and tweaks with it at https://github.com/rq-cluster, but I still have quiet job loss occasionally. Then I reverted back, and found out even rq with redis.py loses job occasionally, i.e. this issue.

**Comment by hitigon** [comment]
@fossilet any ticket you can point to that is related to job loss on rq and redis.py? Also, have you tried the latest redis-py-cluster (1.3.X) ?

**Comment by fossilet** [comment]
@hitigon yes, this issue. I have not tried the new versions. Good luck!

**Comment by hitigon** [comment]
@fossilet which one?

**Comment by jjjjw** [comment]
I have seen this as well using RQ `0.8.0` and python redis `2.10.5`. I am not using redis-py-cluster.

The issue I saw is the same as others reported. I can see jobs in Redis that were never added to the queues, or silently failed in some other way. I don't know how to reproduce, it only happens occasionally. 

My current workaround is to poll Redis for zombie jobs and requeue them.

**Comment by kvnramirez** [comment]
There is any solution for this problem?

**Comment by lifenautjoe** [comment]
Same issue here.

Once in a while we get complaints that somethings not working on the application, 9/10 times its rq workers doing nothing while having thousands of jobs queued.

**Comment by James-Quigley** [comment]
@jjjjw mind sharing what your zombie job requeueing logic looks like?

**Comment by zsblevins** [comment]
Also seeing this issue. I have several applications that use RQ and I've only noticed it in a single application.

The job gets created and persisted in Redis, but the queue doesn't have any reference to the job and it never gets run. Here's some sample behavior:

`
>>> from redis import Redis
>>> rconn = Redis(host="<redacted>", db=5)
>>> from rq.job import Job
>>> job = Job.fetch('f5514216-61cb-4967-a7dc-0a02a43216da', connection=rconn)
>>> job.id
'f5514216-61cb-4967-a7dc-0a02a43216da'
>>> job.get_status()
'queued'
>>> job.origin
'render_v1'
>>> job.requeue() # This shows it's not on the failed job registry and thus can't be requeued
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/apps/python3.6/lib/python3.6/site-packages/rq/job.py", line 529, in requeue
    self.failed_job_registry.requeue(self)
  File "/apps/python3.6/lib/python3.6/site-packages/rq/registry.py", line 194, in requeue
    raise InvalidJobOperation
rq.exceptions.InvalidJobOperation
>>> job.to_dict()
{'created_at': '2019-09-27T21:27:24.600974Z', 'data': b"x\x9c%NK\x0e\x820\x14\xd4\xc4D\xd4\xa5Wp\x81\x0b\x9aBK\x913\xb83\xee\t\x96g@C1\xf4a\xdc\x98x\x80\xee|^\xc1s\xda\xc8\xac\xe6\x93\xcc\xcck\xf6\xf9NF\x84n\x0e\x0f\xd0\x03\x02\xb9\xb5\x05D\xe8\x19\x96\xf6jY\x0f\xa6\x82\x9e\xdc\xf2\xf0'Go\xd2\x9b\xb6O\n]\x00w0X4\x15\xb9\x8dNr)d\x9aE\xb1\x10I$\x13\xa5\xa2\x1d\xf72\xcfN \xa5\x12YuN\xc9-\x1a_\\b\xd3\x19\xdaO]Pw\x16M\xd9\xfa\xcd\x15X\x1e3{\xd1\x9c+\xaf\xac\xae\xa1-\x8b1\x9b\xdd\xc0\x1f\x18\x90\xd8\x0f\x85X<\xae", 'origin': 'render_v1', 'description': "execute(event_id='c2943457-1332-4266-8057-97be44637df5', hostname='<redacted>', iteration=1, schema_name='peer')", 'enqueued_at': '2019-09-27T21:27:24.601033Z', 'timeout': 180, 'failure_ttl': 604800, 'status': 'queued'}
>>> from rq import Queue
>>> queue = Queue('render_v1', connection=rconn)
>>> queue.jobs
[]
>>> queue.fetch_job(job.id)
Job('f5514216-61cb-4967-a7dc-0a02a43216da', enqueued_at=datetime.datetime(2019, 9, 27, 21, 27, 24, 601033))
>>> queue.jobs
[]
>>> queue.failed_job_registry.get_job_ids()
[]
`

**Comment by orionw** [comment]
Currently seeing the same issue.  Any solutions?

**Comment by hasslercastro** [comment]
> Currently seeing the same issue. Any solutions?

Same here

**Comment by mrinalsinha** [comment]
Hello, I'm also facing the same issue, my tasks remain in the queue even when the workers are free. Can anyone please suggest if not this then, what is better to switch to.

**Comment by hasslercastro** [comment]
I was using a docker container and for some reason the workers never took a job out from the queue 
I found this in other thread, basically create workers from code using multiprocessing 

```python
r = redis.Redis(host='redis', port=6379)
q = Queue(connection=r)

#Debug, cleaning old jobs
q.empty()


def need_burst_workers():
    # check database or redis key to determine whether burst worker army is required
    return True #boolean

def num_burst_workers_needed():
    # check the number, maybe divide the number of pending tasks by n
    return 2 #integer


def runworker():
    #Burst mode kills the worker once it has completed its jobs
    qs=["default"]
    with Connection(r):
        if need_burst_workers():
            for i in range(num_burst_workers_needed()):
                multiprocessing.Process(target=Worker(qs).work, kwargs={'burst': False}).start()
        else:
            time.sleep(10) #in seconds

#Workers will read enqueued jobs
runworker()
```

**Comment by nihar4276** [comment]
Hi,
Facing the same issue. Some jobs are never pushed into the queue. Is there any solution?

**Comment by siddharthshah3030** [comment]
I found a workaround 
if a job is stuck at "queued", i automatically delete and enqueue again

```python
from rq import Queue
from redis import Redis
redis_conn = Redis()
q = Queue(connection=redis_conn)

job = q.fetch_job('my_id') # Returns job having ID "my_id"
job.get_status() # -> 'queued'

job.delete() # -> deletes the job 

job.get_status() # -> "" job does not exist

# now enqueue job again normally and it works and runs successfully
```

**Comment by atainter** [comment]
For what it's worth, my team is having the same problem. It's making RQ unusable for us. Seems like this is a bug.

**Comment by selwin** [comment]
@atainter are you able to nail down whether the job was actually ever in the queue (pushed to Redis list)?

I’m just curious because a number of people are reporting this bug, but none are able to reproduce this. Perhaps it’s a temporarily blip with the worker box or something similar?

**Comment by atainter** [comment]
Yes, looks like the job is in in Redis and stuck in a "queued" status. I'm not able to reproduce with a certain job - it randomly happens, but it's somewhat frequent. I looked for zombie workers, but everything seemed normal.

**Comment by selwin** [comment]
In that case, let me work on a patch that increases debug logging around this area so we can try to nail down what happens the next time this happens again.

**Comment by atainter** [comment]
Hey @selwin any update on this?

**Comment by atainter** [comment]
After some more digging, my theory is that the job is added to Redis as `rq:job:ID`, but not added to 'rq.default'. This would explain why `queue.jobs` does not fetch the job, but `job_class.fetch(id)` does. One reads from the 'rq:job' key while the other gets ids from 'rq:default' and uses them to fetch data from 'rq:job'. 

I'm going to continue to try to look into this. Maybe there's silent redis connection error when RQ enqueues the job in the 'rq:default' key?

**Comment by atainter** [comment]
Under further investigation, it looks like this enqueue uses a pipeline to update redis. I wonder if one if the commands is silently failing: https://stackoverflow.com/a/54153873

**Comment by selwin** [comment]
@atainter sorry I've been super busy this week.

> Under further investigation, it looks like enqueue uses a pipeline to update redis. I wonder if one if the commands is silently failing

I don't think this is the problem. My first guess would be the job is dropped after the job is popped of the queue (Redis list), but before the job is put on the `StartedJobRegistry` (sorted set).

My intention is to add `logging.debug()` calls around this area during the enqueue process:
1. Job is successfully added to queue (Redis list)

And on the worker process, add `logging.debug()` here:
1. Prior/after job is popped of the queue
2. Prior/after job is added to StartedJobRegistry

I think the job became orphaned between steps 1 & 2.

Having said that, I've never had these kind of issues though, could it be that the hardware you're running RQ on is overutilized?

**Comment by atainter** [comment]
I don't think it's underutilized, but I could be wrong. We're in Kubernetes and running 1000m (reqs) to 1500m (limit). Our Redis instance is hosted in AWS. Not sure if there could be some cross domain issue there, but it works in other environments. I'll look into this with some monitoring tools.

I added some code in a subclass to log prior/after job is popped of the queue. Let me see if I can gather any data from that and I'll look into the StartedJobRegistry in parallel.

**Comment by atainter** [comment]
Okay, so I was able to reproduce with some logging. It looks like jobs that get stuck in this queued state call 'rpush' to add the job to the Redis queue, but they don't get popped off the queue in the worker. i.e. this code doesn't get called for stuck jobs, but it does for other jobs:

```
class QueueWithLogging(Queue):
    @classmethod
    def lpop(cls, queue_keys, timeout, connection=None):
        result = RQQueue.lpop(queue_keys, timeout, connection=connection)

        if result is not None:
            queue_key, job_id = map(as_text, result)
            queue_logger.info(f"Popping job_id from the queue: {queue_key} {job_id}")

        return result
```

**Comment by selwin** [comment]
But were you able to verify that the job successfully get pushed into the list?

One way to find out about this:
1. Stop your workers
2. Put a logging call after `rpush` and confirm that the job ID is indeed in the list after the job is enqueued

Repeat this for a few hundred/thousand jobs and see if any of them actually don't get pushed to the list.

**Comment by atainter** [comment]
Okay, I just did a bunch of testing for this and I think I found something. Originally I had 6 workers and this was happening quite often. I scaled back to 1 worker and I had a really difficult time reproducing this, so I ramped back up to 2. To figure out what was going on, I logged the current state of the job_id queue in the service after all jobs in a batch had been scheduled AND in a worker before it calls the blpop method.

Here's a spreadsheet of the logs from my machines:
https://docs.google.com/spreadsheets/d/1dbh1NokDO-pjkwnzDjRzWU1kuhS9EgnMgrbwSeCUY3s/edit?usp=sharing

Some things to note: 
1. Logs are in reverse chronological order with oldest logs at the bottom
2. I noticed this batch has some duplicate jobs due to a bug in our code. I don't think this would matter, but I'm not sure
3. The lost job ID is bolded in red

Looks like the workers blpop jobs off the queue right as the queue writes to it using an rpush. Workers 1 and 2 pop 
`ffd2d819-43d7-4c60-8654-c23af2fe4ab3` and `b177150f-4f53-442a-a734-f053c3dab1b8` off the queue right as  `4fd996cd-a4f6-4a09-a4ce-0b877acff2d9` is pushed into the queue. If you follow the logs, you'll see that `4fd996cd-a4f6-4a09-a4ce-0b877acff2d9` is never popped and is not in the job queue when the service logs the ids.

This leads me to believe that there is a **race condition** that happens between blpop and rpush. Is there a way to verify this? Can lock the table to prevent this from happening? I'm not a Redis expert.

**Comment by atainter** [comment]
The other interesting thing I found is that RedisPy swallows a ConnectionError for a `pipeline.execute` call: https://github.com/andymccurdy/redis-py/blob/master/redis/client.py#L4156-L4169

Looks like BusyLoadingError also extend ConnectionError: https://github.com/andymccurdy/redis-py/blob/master/redis/exceptions.py#L20

I'm not sure what BusyLoadingError or ConnectionError is, but I wonder if this could also be silently failing because the blpop is popping the queue during the rpush and the key is "busy".

**Comment by atainter** [comment]
@selwin I was able prevent data loss by locking the workers before and after adding jobs to the queue. It seems like this only happens if I send more jobs at one time that there are workers. I think the race condition happens when the last job_id is popped off the queue and a job is added at the same time. 

Here's my pseudo code:

```python
    try:
        suspend(redis_conn)

        # Hack to break workers out of blpop
        workers = Worker.all(redis_conn)
        for _ in workers:
            job_queue.enqueue(noop_job, result_ttl=60)

        for txn in txns:
            job_queue.enqueue(job_func, txn.data, job_id=txn.id)
    
    finally:
        resume(redis_conn)
```

This doesn't seem like a long term fix, though. Jobs could be added from several services and there could be a race condition that resumes the workers in one server right after they're suspended in another. Additionally, there are a bunch of noop jobs that are unnecessarily added. 

Perhaps you could switch to this pattern instead? https://redis.io/commands/rpoplpush#pattern-reliable-queue

**Comment by selwin** [comment]
> This leads me to believe that there is a race condition that happens between blpop and rpush. Is there a way to verify this? Can lock the table to prevent this from happening? I'm not a Redis expert.

Redis is single threaded by design so race conditions can't happen within Redis itself, only within our code. 

> I think the race condition happens when the last job_id is popped off the queue and a job is added at the same time.

So this can't actually happen.

Reading your logs, it seems like job `4fd996cd-a4f6-4a09-a4ce-0b877acff2d9` was never pushed into the queue. If this is the case, using `rpoplpush` won't help, since it was never in the queue (list) to begin with. And this command works doesn't work for multiple keys.

So I think what happens is when `rpush` is called, Redis is busy so it silently drops it. What I'll do is see whether we can check the result of `pipeline.execute()` to see whether we can detect this. Do you mind logging the result of `pipeline.execute()` here https://github.com/rq/rq/blob/master/rq/queue.py#L481 ?

Sorry for the hassle but I can't reproduce this on my setup.

**Comment by atainter** [comment]
@selwin I made a change to add additional logging and merged into the env where we were seeing this issue. However, right before merging the change, we decided to update our Redis cluster from **5.0.4** to **5.0.6** [release notes](https://raw.githubusercontent.com/redis/redis/5.0/00-RELEASENOTES).

I disabled my temporary workaround and haven't been able to reproduce 🤞 . I'll continue monitoring with my logging changes for the next couple days and I'll let you know if anything changes. Thanks for the help!

**Comment by hossein761** [comment]
I removed celery from my project to use a "simpler" tool, but having the same issue. my Redis version 5.0.10. Using django, I see the jobs in the default queue with the status: scheduled they are stuck in there.

**Comment by jtfidje** [comment]
Related to this, we now experience an issue where a job gets stuck in the `deferred` state. occasionally when a job depends on another, it won't get queued when the first job finishes. I can't reproduce it by "will", but if I queue the same set of jobs enough times it always happens eventually.

**Comment by LorenRd** [comment]
Just opened this issue https://github.com/rq/rq/issues/1433 Can someone help?

**Comment by zsblevins** [comment]
An additional bit of context to add. We hit this bug again in another application when we switched from the default Worker to SimpleWorker (to workaround issues with GRPC streaming being incompatible with forking). It occurs VERY frequently with the move to SimpleWorker.

**Comment by abhisuri97** [comment]
I came across this issue as well. Killing the worker thread and restarting seems to "fix" it, but I'd rather not have to resort to that in production. Does anyone have any idea what to do here?

**Comment by essadek** [comment]
I'm having the same issue, I noticed that when jobs are chained one after one immediately the jobs are not executed, however if there is some interval in between the jobs get executed.

**Comment by daverin** [comment]
same findings as @abhisuri97 in production. I've spotted it a few times locally in my docker compose environment, but thus far I haven't been able to consistently reproduce the issue.

**Comment by dpereira** [comment]
This report looks eerily familiar:

https://github.com/redis/redis/issues/7128

We had this issue in production, and went through the following changes:

 - Removed rq in favor of an in-house client lib for queueing/dequeing messages.
 - Changed redis client implementation
 - Finally, replaced redis and the whole implementation with one based in Active MQ.
 
 We only stopped seeing behaviors like the ones in this thread after we stopped using redis for queue/jobs entirely.

**Comment by selwin** [comment]
So it’s a Redis issue? If it is, then unfortunately there’s nothing that we can do.

**Comment by dpereira** [comment]
I can't say 100% it was redis. Can't say it isn't.

It could be a bug related to incorrect redis usage in rq that we carried in our implementation as well, for example. Just wanted to share our experience, in the hope this helps with the troubleshooting.

**Comment by viorelspinu** [comment]
Experienced the same behaviour. For me it was running multiple Redis servers instance locally with Docker and jobs were randomly published to different servers. Only one server had workers attached to it. After stopping all other Redis servers and keeping up only the one with workers connected this issue did not happen anymore.

**Comment by gabriels1234** [comment]
Hi,
Any update on the original issue?

**Comment by gaffney** [comment]
We have used and loved RQ for 4 years but now that our company is experiencing heavier loads and using background jobs more extensively we are running into this issue quite often. Unfortunately this is a deal breaker as this is turning into a near-constant on-call headache.

Happy to help debug in the meantime as I won't be migrating the code for at least a few weeks.

Also worth mentioning, we are using the latest versions of RQ & Redis client, so the latest changes I see merged above have not fixed this 100%.

**Comment by selwin** [comment]
@gaffney do you mind confirming this?

1. When the jobs don't run, do they have complete data (all keys in hashes are present)?
2. Do the job keys have TTL assigned to them? (maybe they have been executed but not marked as such)
3. Are the job keys present in the queue (Redis list)? The queue keys are prefixed with `rq:queue` https://github.com/rq/rq/blob/master/rq/queue.py#L42

**Comment by baslia** [comment]
Same issue here, randomly the queue stopped working.

**Comment by y09esh** [comment]
Same issue i am also facing. before upgrading packages everything was working fine. it seems that latest versions of redis and rq not supporting each other. can anyone confirm which versions are working fine?

**Comment by lowercase00** [comment]
@baslia and @y09esh can you provide more information on the scenarios, versions and conditions you've encountered this issue? this is important to try to replicate and fix it.

**Comment by alekh1stzoom** [comment]
In my case the jobs were getting enqueued, as well there was a worker, but still no jobs were getting executed, had to start another worker manually, which executed all the pending tasks. This is weird.

**Comment by csangonzo** [comment]
> In my case the jobs were getting enqueued, as well there was a worker, but still no jobs were getting executed, had to start another worker manually, which executed all the pending tasks. This is weird.

Same here, makes my service completely unreliable

**Comment by lowercase00** [comment]
@alekh1stzoom @csangonzo  any other details you guys can share? haven't been able to reproduce a scenario where workers suddenly stop to work. Is the worker process running? Is the job being sent to the queue? Any TTL as @selwin asked? are the jobs sync/async? custom queue/classes? single redis server or cluster, sentinel?

**Comment by alekh1stzoom** [comment]
Even I was not able to reproduce it. I was running it inside a docker and the worker process was running and the state was 'busy'. The jobs were being send to the queue and hence the queue size was growing rapidly in my case. My jobs were sync. I had created my own custom queue on a single redis server. Here is the ss attached.

![Screenshot from 2022-09-14 22-50-20](https://user-images.githubusercontent.com/66582764/192932775-13bdf163-d5b3-4eb2-8ea7-fa23208d982a.png)

**Comment by rizwan777** [comment]
i think need to move forward to pick Cerlery bit.

**Comment by ericman93** [comment]
I have the same issue. if I'm enqueuing more tasks than the number of workers, sometimes one job isn't starting
I do see logs from `execute_job`, but not from `prepare_for_execution`

answering @selwin questions
> When the jobs don't run, do they have complete data (all keys in hashes are present)?

yes

> Do the job keys have TTL assigned to them? (maybe they have been executed but not marked as such)

no

> Are the job keys present in the queue (Redis list)? The queue keys are prefixed with rq:queue

the prefix of the job is `rq:job:`

and @lowercase00 
> are the jobs sync/async?

async

> custom queue/classes? 

yes, for both (but 99% of the time everything is working)

> single redis server or cluster, sentinel?

sentinel

**Comment by lowercase00** [comment]
@selwin wondering how to test this. I personally haven't experienced this, but I do refresh workers, so maybe not the best benchmark. There doesn't seem to be any explicit pattern on this issue, and for all we know can't be replicated.

I started a test that sends 200 tasks per second to the queue, 4 workers running. I'll leave it running for a few hours (maybe days?) to see if something happens. I can probably compare de number of jobs sent with jobs finished at the end. Any other idea on how to try to replicate?

It would be possible to have a workaround that requeues dead jobs (queued for X amount of time), or maybe one that refreshes the worker process, don't know...

**Comment by selwin** [comment]
I'm actually running out of ideas to debug this because:
1. It doesn't look like there's any bug related to this in regular use
2. There's anecdotal evidence the issue actually is only resolved after [ditching Redis entirely ](https://github.com/rq/rq/issues/758#issuecomment-833207507)

The next steps that I'd try are:
1. Use debug logging (I just checked logging statements around the dequeueing logic and I think we may need to add more logging statements here)
2. Try running `Worker` with lower TTL. `Worker(default_worker_ttl=120)`

**Comment by lowercase00** [comment]
Yeah... I made a couple of tests:
- ~200 tasks per second, 4 workers, up until around 500k tasks finished, no tasks lost.
- 10 tasks per second, single worker, just hit 100k tasks, no tasks lost.

Debug logging should help, specially around the `Queue.lpop()`, I can make a quick PR if you want.
Current Worker TTL is set to 420, it might be an option to reduce the default to 120 maybe?
It's not entirely clear to me the heartbeat/worker_ttl logic, but I did notice that `rq info` might not show workers sometimes (when they are actually working), which is not ideal. It might be something to investigate further.

The only thing I can think of is the potential issues related Redis. There is one topic in `redis-py` [here](https://github.com/redis/redis-py/issues/1305#issuecomment-597305775) that is related to the possible `blpop` blocking forever issue, which was mentioned [here](https://github.com/redis/redis/issues/7128) and [here]([url](https://github.com/redis/redis-py/issues/1331)), it does seem like a good approach: have the socket timeout as a safeguard for any issues with `blpop` hanging. The only place I could see the `socket_timeout` being used in the codebase is at the Sentinel config, so this is not set by default.

Since the Redis object is being passed to the queue, maybe the queue could set the `socket_timeout` when initializing

```python
from redis import Redis
r = Redis()
r.config_set("socket_timeout", 30)
```

A more complex (and robust) logic could include checking for the `timeout` being used at the `dequeue_any`, no sure what the downside would be since there would be a "risk" of changing the configuration multiple times.

Do you see any downsides of this approach?

**Comment by selwin** [comment]
> Debug logging should help, specially around the Queue.lpop(), I can make a quick PR if you want.

Yes please.

> Since the Redis object is being passed to the queue, maybe the queue could set the socket_timeout when initializing

Sounds good, we can set `socket_timeout` to `default_worker_ttl + 5` on worker `__init__`.

**Comment by lowercase00** [comment]
Logs are [here](https://github.com/rq/rq/pull/1752).
Suggestion on dynamic connection configuration (with both retry logic and a dynamic timeout for the socket) [here](https://github.com/rq/rq/pull/1753)

**Comment by lowercase00** [comment]
For reference, to increase guarantees, this can also be interesting: https://mrq.readthedocs.io/en/latest/jobs-maintenance/.
We could add a `RequeueStaledJobs` maintenance job. That, the `socket_timeout` with retry logic in the redis connection, and a simple retry logic on the `work()` method, should decrease the odds of losing jobs significantly.

**Comment by gabriels1234** [comment]
Naive question:
I know that the R in RQ stands for Redis.

What are the true benefits of letting Redis handle it, and not, for example, Postgres?

If latency is the issue, I don't see how a few ms extra would be so important for jobs that  are mostly ok with sub-second performance?

**Comment by gabriels1234** [comment]
@lowercase00 I'm just bringing this up here because there's a reliability issue that's even prompting and making @selwin consider ditching Redis. I'm just asking whether using Postgres or any other SQL db would bring integrity, reliability while costing a few more ms.

**Comment by lowercase00** [comment]
@gabriels1234 understood your point. The quick answer is no.

Now the long answer: not necessarily changing the datastore will bring reliability/integrity and the tradeoff involves a lot more than just "a few more ms". You can make perfectly reliable working queues in Redis, just as you can in many other backends (even SQL-based ones, that were definitely not made for queues) - the so called "integrity" in SQL-based datastore has little to do with the "reliability" we are talking about in this issue. You will, however, encounter a series of other challenges (connection handling, locks, schemas, migrations, backups, performance, etc etc) - reliability is not exclusively dependent on the backend as it may appear so at first sight. Implementation is key (whichever the backend), and this is no different with Redis, RabbitMQ, ActiveMQ, SQS, Postgres, Mongo, ZeroMQ, Kafka or any other backend for that matter - each have different characteristics and work differently. All of those are production-ready and battle tested, which does not mean that you won't have issues - it all depends on how you use then.

[This page](https://www.rabbitmq.com/reliability.html) on RabbitMQ docs can also help get a grasp on the many ways in which a messaging system may have reliability issues - several of the points raised by RabbitMQ's docs can be extrapolated to tasks queue systems. That's an interesting read, and should give you an idea on how and why reliability is not that straight forward. Postgres also has a [dedicated page](https://www.postgresql.org/docs/current/wal-reliability.html) on reliability, interesting read as well.

@selwin knows the codebase way better than myself, but I personally I didn't read his comment as a consideration to ditch Redis, but rather that this may be an issue with Redis itself (and not RQ). From what we've seen so far, it may well be an issue with a specific Redis command (`blpop`), but again this is just a hypothesis. Following the links on previous comments, you will see discussions on Redis boards that also acknowledge the possibility of `blpop` hanging, and also a suggestion to introduce a timeout mechanism at the socket level as a safety net.

All of this to say: even if there is a specific Redis behavior that for now is affecting RQ's reliability, this can and will be fixed, and changing the datastore doesn't seem like a solution (specially since we just don't know if Redis is actually to problem at this point), is definitely not an easy feat to get right (if you ever worked with Celery + RabbitMQ you probably know the huge amount of configuration that you can get wrong and can lead to job lost, I've been there too many times). And that's definitely not to say that RabbitMQ's is "unreliable" just as Redis isn't.

If you do experience a reliability issue, would be great if you could share your setup in a way that could help us replicate (so far this issue hasn't been reliably replicated), troubleshoot and fix this issue as soon as possible.

**Comment by gabriels1234** [comment]
@lowercase00 I came back to this issue because I'm trying to figure out why a scheduled job in the past is being kept in the scheduled registry while newer ones are finally getting enqueued. 
I came across code in scheduler.py (lines 150-166) that pointed to: 
"#TODO: Try to use Lua script to make get_jobs_to_schedule() and remove_jobs() atomic"

That atomicity is easily achievable using an SQL transaction. 

All the other reliability concerns apply to everything.
Just wanted to clarify that point.

**Comment by lowercase00** [comment]
@gabriels1234 do you have a code sample where this issue can be replicated?
@gaffney if you're still running `rq`, would you mind trying to set a `socket_timeout` to your Redis connection (anything higher than the `worker_default_ttl` which is 420 by default should be enough) to check whether the issue persists? Also, some logging around the `lpop` method would also be helpful to assess whether this is where the job went missing.

**Comment by selwin** [comment]
> What are the true benefits of letting Redis handle it, and not, for example, Postgres?

Redis has native data structures that's suitable for implementing task queues.

Yes, in theory you can use Postgres for everything, from relational database, document based storage using JSONB or even as a pure key value store. Some people have indeed built task queues based on Postgres. Some things will be easier to implement on Redis, some will be easier to implement on Postgres.

I personally use Postgres for main datastore at work and Redis for caching, task queue and atomic locks and increments.

**Comment by ericman93** [comment]
`socket_timeout` solved my issue

**Comment by lowercase00** [comment]
@ericman93 great news - this is super critical, glad to know it helped.
Now I couldn't really find anything on Redis Py as to why `BLPOP` would hang.
Next release should have the `socket_timeout` being set by RQ (if it was not set by the user), so hopefully this will solve the issue - at least making the worker fail explicitly.
Let us know if you find anything else.

**Comment by stv8** [comment]
Chiming in that I am currently experiencing this issue on `rq v1.16.2` and overriding the socket timeout of the worker to 500 seconds didn't help, we still see jobs that are somehow being popped off the queue but not worked on.

I describe some peculiarities about those values here https://github.com/rq/rq/issues/2099
